### PR TITLE
Add networkx layout to GraphVisual

### PR DESCRIPTION
**Pull request by cvanelteren**
Added a Quality of Life improvement allowing for any arbitrary position coordinates of nodes in the graph. Allows for interfacing with networkx layouts by string, or dictionary input. Alternatively one could give it a numpy array for plotting. 

An alternative would be to make the networkx import optional, leaving the numpy array as default. 

Example:

```python
from vispy import app, scene
from vispy.visuals.graphs import layouts
import networkx as nx
g = nx.erdos_renyi_graph(100, .2)
layout = layouts.get_layout("networkx_layout", graph = g, layout = "spring")

# create canvas
canvas = scene.SceneCanvas(title='Simple NetworkX Graph', size=(600, 600),
                           bgcolor='black', show=True)
view = canvas.central_widget.add_view('panzoom')
visual = scene.visuals.Graph(
    layout.adj, layout = layout, line_color= (1, 1, 1,  .5), arrow_type="stealth",
    arrow_size=30, node_symbol="disc", node_size=20,
    face_color=(1, 0, 0, 1), border_width=0.0, animate=False, directed=False,
    parent=view.scene)


@canvas.events.draw.connect
def on_draw(event):
    if not visual.animate_layout():
        canvas.update()
    
canvas.show()
```

### CODE REVIEW

File: vispy/visuals/graphs/layouts/networkx_layout.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    def __init__(self, graph, layout = None, *args, **kwargs):
+        self.graph = graph
+        self.positions = np.zeros((len(graph), 2), dtype = np.float32)
+        # default random positions
+        if type(layout) is type(None):
+            self.positions = np.random.rand(*self.positions.shape)
+
+        # check for networkx
+        elif isinstance(layout, str):
+            if USE_NETWORKX := importlib.util.find_spec("networkx"):
```
Comment (djhoese) [inline vispy/visuals/graphs/layouts/networkx_layout.py:16]: We still support python 3.6+, the walrus operator was introduced in python 3.8. Could you rework this line?

Comment (djhoese) [reply vispy/visuals/graphs/layouts/networkx_layout.py:16]: Maybe just a try/except at the top of the module? That's how we normally do this type of thing in the other vispy modules.

**Review by djhoese — CHANGES_REQUESTED** [review]
Very interesting. I don't really use the graphs myself so I'm not sure how much feedback I can provide on the actual concepts used, but it looks really useful. Do you think this is useful enough that your example code should be added to vispy's `examples/`?

There is one issue with your walrus operator which I've made an inline comment about. Otherwise, there are a few styling issues that might get flagged now or maybe later. Do you think you could add some docstrings to the NetworkxCoordinates class? And remove the spaces on your keyword argument default parameters (`layout = None` -> `layout=None`)?

Thanks a lot for this.

File: vispy/visuals/graphs/layouts/networkx_layout.py (hunk 1 above)
Comment (cvanelteren) [reply vispy/visuals/graphs/layouts/networkx_layout.py]: ah my bad in my haste I haven't checked if there is a developers manual. I use networks quite a lot and have used vispy in the past for another project and was pretty impressed with the performance. I have edited the code slightly to do a try/except import. Alternatively we could use `importlib` for this find `util.find_spec` being the default way of looking for packages.

Comment (cvanelteren) [reply vispy/visuals/graphs/layouts/networkx_layout.py]: The latest commit has
- fixed spec according to python 3.6; removed walrus operator
- added try except import
- added doc string

**Comment by djhoese** [comment]
Thanks for all the fixes. There isn't currently a developer's manual, but is something I'm working on. It looks like travis is failing due to some flake8 issues:

```
Running flake8... 

vispy/visuals/graphs/layouts/networkx_layout.py:6:1: E722 do not use bare 'except'

vispy/visuals/graphs/layouts/networkx_layout.py:11:1: E302 expected 2 blank lines, found 1

vispy/visuals/graphs/layouts/networkx_layout.py:32:57: E251 unexpected spaces around keyword / parameter equals

vispy/visuals/graphs/layouts/networkx_layout.py:32:59: E251 unexpected spaces around keyword / parameter equals

vispy/visuals/graphs/layouts/networkx_layout.py:34:25: E721 do not compare types, use 'isinstance()'

vispy/visuals/graphs/layouts/networkx_layout.py:41:40: E261 at least two spaces before inline comment

vispy/visuals/graphs/layouts/networkx_layout.py:65:47: E251 unexpected spaces around keyword / parameter equals

vispy/visuals/graphs/layouts/networkx_layout.py:65:49: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:1:1: F401 'vispy.app' imported but unused

examples/basics/visuals/networkx_layout_example.py:5:53: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:5:55: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:5:65: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:5:67: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:12:23: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:12:25: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:12:45: E251 unexpected spaces around keyword / parameter equals

Failed: flake8 failed
```

Additionally, would it be possible for you to add some tests for this? And lastly, could you copy the header from the other example scripts and put it on your example? Thanks!

**Comment by cvanelteren** [comment]
> Thanks for all the fixes. There isn't currently a developer's manual, but is something I'm working on. It looks like travis is failing due to some flake8 issues:
> 
> ```
> Running flake8... 
> 
> vispy/visuals/graphs/layouts/networkx_layout.py:6:1: E722 do not use bare 'except'
> 
> vispy/visuals/graphs/layouts/networkx_layout.py:11:1: E302 expected 2 blank lines, found 1
> 
> vispy/visuals/graphs/layouts/networkx_layout.py:32:57: E251 unexpected spaces around keyword / parameter equals
> 
> vispy/visuals/graphs/layouts/networkx_layout.py:32:59: E251 unexpected spaces around keyword / parameter equals
> 
> vispy/visuals/graphs/layouts/networkx_layout.py:34:25: E721 do not compare types, use 'isinstance()'
> 
> vispy/visuals/graphs/layouts/networkx_layout.py:41:40: E261 at least two spaces before inline comment
> 
> vispy/visuals/graphs/layouts/networkx_layout.py:65:47: E251 unexpected spaces around keyword / parameter equals
> 
> vispy/visuals/graphs/layouts/networkx_layout.py:65:49: E251 unexpected spaces around keyword / parameter equals
> 
> examples/basics/visuals/networkx_layout_example.py:1:1: F401 'vispy.app' imported but unused
> 
> examples/basics/visuals/networkx_layout_example.py:5:53: E251 unexpected spaces around keyword / parameter equals
> 
> examples/basics/visuals/networkx_layout_example.py:5:55: E251 unexpected spaces around keyword / parameter equals
> 
> examples/basics/visuals/networkx_layout_example.py:5:65: E251 unexpected spaces around keyword / parameter equals
> 
> examples/basics/visuals/networkx_layout_example.py:5:67: E251 unexpected spaces around keyword / parameter equals
> 
> examples/basics/visuals/networkx_layout_example.py:12:23: E251 unexpected spaces around keyword / parameter equals
> 
> examples/basics/visuals/networkx_layout_example.py:12:25: E251 unexpected spaces around keyword / parameter equals
> 
> examples/basics/visuals/networkx_layout_example.py:12:45: E251 unexpected spaces around keyword / parameter equals
> 
> Failed: flake8 failed
> ```
> 
> Additionally, would it be possible for you to add some tests for this? And lastly, could you copy the header from the other example scripts and put it on your example? Thanks!

As there is no development manual, could you provide me some pointers to what standards you are using? From the docstrings I deduced the numpy standard is being used. The `testing` subfolders contains no unittest framework, is it okay if `nose` is used or just simple functions? Then I can start converting the code to `flake8`. Thanks in advance!

**Comment by djhoese** [comment]
We semi-recently switched to requiring pytest so I would say feel free to use any builtin pytest functionality. Due to the way we have to test multiple backends, our testing initialization is a little difficult to understand at first glance, but the individual test modules are pretty plain pytest + some decorators to say what is required to run the test (need a GL canvas, etc). You should be able to run your specific tests with your normal pytest invocation `pytest path/to/your/test_module.py`.

Docstrings are numpy style, yes. They aren't being as heavily validated as they should be but that's another thing I'm working on in another pull request. Otherwise, right now if the code passes with flake8 then the code is good enough. Docstrings can be minimally tested by installing `flake8-docstring` and running flake8 on your module. This will likely fail for a lot of vispy and there is a hacktoberfest issue for that also. The rest of this year is basically "revamp all of vispy" time.

File: vispy/testing/tests/test_networkx_layout.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#format pep8
```
Comment (djhoese) [inline vispy/testing/tests/test_networkx_layout.py:1]: Would you mind adding the header to the top of this to match the other modules in vispy (at least most of them):

```
# -*- coding: utf-8 -*-
# Copyright (c) Vispy Development Team. All Rights Reserved.
# Distributed under the (new) BSD License. See LICENSE.txt for more info.
```

File: vispy/visuals/graphs/layouts/networkx_layout.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        Parameters
+        ----------
+        graph : a networkx graph.
+        layout : str or dict or iterable-object of float32, optional
+        - When :layout: is s string, a lookup will be performed in the networkx avaiable layouts.
+        - When :layout: is a dict, it will be assumed that it takes the shape (key, value) = (node_id, 2D-coordinate).
+        - When :layout: is numpy array it is assumed it takes the shape (number_of_nodes, 2).
+
+
+        Yields
```
Comment (djhoese) [inline vispy/visuals/graphs/layouts/networkx_layout.py:27]: I've never seen `Yields` as part of a numpy docstring. I could understand it for a generator, but is this expected for an `__init__` method? Shouldn't this be `Returns` on the `__call__` method?

File: examples/basics/visuals/networkx_layout_example.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+from vispy import app, scene
```
Comment (djhoese) [inline examples/basics/visuals/networkx_layout_example.py:1]: Header here too. Thanks.

Comment (cvanelteren) [reply examples/basics/visuals/networkx_layout_example.py:4]: done

File: vispy/visuals/graphs/layouts/networkx_layout.py (hunk 3 above)
Comment (cvanelteren) [reply vispy/visuals/graphs/layouts/networkx_layout.py]: done

File: vispy/testing/tests/test_networkx_layout.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) Vispy Development Team. All Rights Reserved.
+# Distributed under the (new) BSD License. See LICENSE.txt for more info.
+from vispy import testing
+from vispy.visuals.graph.layouts import get_layout
```
Comment (djhoese) [inline vispy/testing/tests/test_networkx_layout.py:5]: `graphs`? The tests are failing on this import.

File: vispy/testing/tests/test_networkx_layout.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        testing.assert_true(
+            isinstance(layout(graph, layout="circular"), layout))
+
+        # testing dict input
+        pos = nx.circular_layout(graph)
+        testing.assert_true(isinstance(layout(graph, pos), layout))
+
+    else:
+        # testing no graph
+        testing.assert_raises(ValueError, layout(graph, positions))
```
Comment (djhoese) [inline vispy/testing/tests/test_networkx_layout.py:43]: Should this test happen regardless of the current environment? Maybe mock the networkx dependency in a separate test?

**Review by djhoese — COMMENTED** [review]
What do you think about skipping the tests that require networkx if it isn't installed?

File: vispy/testing/tests/test_networkx_layout.py (hunk 6 above)
Comment (cvanelteren) [reply vispy/testing/tests/test_networkx_layout.py]: Apologies. I quickly wrote the tests at work and forgot to test them. The recent commit comes out clean with `pytests`

**Comment by djhoese** [comment]
Looks like there are still some flake8 issues:

```
vispy/visuals/graphs/layouts/networkx_layout.py:8:1: E722 do not use bare 'except'

vispy/visuals/graphs/layouts/networkx_layout.py:13:1: E302 expected 2 blank lines, found 1

vispy/visuals/graphs/layouts/networkx_layout.py:14:29: E251 unexpected spaces around keyword / parameter equals

vispy/visuals/graphs/layouts/networkx_layout.py:14:31: E251 unexpected spaces around keyword / parameter equals

vispy/visuals/graphs/layouts/networkx_layout.py:32:57: E251 unexpected spaces around keyword / parameter equals

vispy/visuals/graphs/layouts/networkx_layout.py:32:59: E251 unexpected spaces around keyword / parameter equals

vispy/visuals/graphs/layouts/networkx_layout.py:34:25: E721 do not compare types, use 'isinstance()'

vispy/visuals/graphs/layouts/networkx_layout.py:41:40: E261 at least two spaces before inline comment

vispy/visuals/graphs/layouts/networkx_layout.py:65:47: E251 unexpected spaces around keyword / parameter equals

vispy/visuals/graphs/layouts/networkx_layout.py:65:49: E251 unexpected spaces around keyword / parameter equals

vispy/testing/tests/test_networkx_layout.py:12:1: E722 do not use bare 'except'

examples/basics/visuals/networkx_layout_example.py:4:1: F401 'vispy.app' imported but unused

examples/basics/visuals/networkx_layout_example.py:8:53: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:8:55: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:8:65: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:8:67: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:15:23: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:15:25: E251 unexpected spaces around keyword / parameter equals

examples/basics/visuals/networkx_layout_example.py:15:45: E251 unexpected spaces around keyword / parameter equals
```

**Comment by djhoese** [comment]
A few left:

```
vispy/visuals/graphs/layouts/networkx_layout.py:8:1: E722 do not use bare 'except'

vispy/visuals/graphs/layouts/networkx_layout.py:36:25: E721 do not compare types, use 'isinstance()'

vispy/testing/tests/test_networkx_layout.py:12:1: E722 do not use bare 'except'

examples/basics/visuals/networkx_layout_example.py:4:1: F401 'vispy.app' imported but unused
```

**Review by djhoese — APPROVED** [review]

**Review by djhoese — COMMENTED** [review]
Sorry, one more change, could you remove the `_example` from your example script's filename?

**Comment by djhoese** [comment]
Some how missed this last commit. This all looks good. Thanks again for all the work and patience!

**Pull request merged**